eth2_testgen/bls/tgen_bls.py

```diff
@@ -0,0 +1,188 @@
+]
+
+PRIVKEYS = [
+    # Private keys are 48 bytes, Keccack hashing only produces 32 bytes
+    # So we concat two and slice
+    big_endian_to_int((hash_eth2(b'Alice') + hash_eth2(b'ecilA'))[:48]),
+    big_endian_to_int((hash_eth2(b'Bob') + hash_eth2(b'boB'))[:48]),
+    big_endian_to_int((hash_eth2(b'Eve') + hash_eth2(b'evE'))[:48])
+]
+
+
+def hash_message(msg: bytes,
+                 domain: int) ->Tuple[Tuple[str, str], Tuple[str, str], Tuple[str, str]]:
+    """
+    Hash message
+    Input:
+        - Message as bytes
+        - domain as uint64
+    Output:
+        - Message hash as a G2 point
+    """
+    return [
+        [
+            int_to_hex(fq2.coeffs[0]),
+            int_to_hex(fq2.coeffs[1]),
+        ]
+        for fq2 in bls.hash_to_G2(msg, domain)
+    ]
+
+
+def hash_message_compressed(msg: bytes, domain: int) -> Tuple[str, str]:
+    """
+    Hash message
+    Input:
+        - Message as bytes
+        - domain as uint64
+    Output:
+        - Message hash as a compressed G2 point
+    """
+    z1, z2 = bls.compress_G2(bls.hash_to_G2(msg, domain))
+    return [int_to_hex(z1), int_to_hex(z2)]
+
+
+if __name__ == '__main__':
+
+    # Order not preserved - https://github.com/yaml/pyyaml/issues/110
+    metadata = {
+        'title': 'BLS signature and aggregation tests',
+        'summary': 'Test vectors for BLS signature',
+        'test_suite': 'bls',
+        'fork': 'tchaikovsky',
+        'version': 1.0
+    }
+
+    case01_message_hash_G2_uncompressed = []
+    for msg in MESSAGES:
+        for domain in DOMAINS:
+            case01_message_hash_G2_uncompressed.append({
+                'input': {'message': '0x' + msg.hex(), 'domain': int_to_hex(domain)},
+                'output': hash_message(msg, domain)
+            })
+
+    case02_message_hash_G2_compressed = []
+    for msg in MESSAGES:
+        for domain in DOMAINS:
+            case02_message_hash_G2_compressed.append({
+                'input': {'message': '0x' + msg.hex(), 'domain': int_to_hex(domain)},
+                'output': hash_message_compressed(msg, domain)
+            })
+
+    case03_private_to_public_key = []
+    #  Used in later cases
+    pubkeys = [bls.privtopub(privkey) for privkey in PRIVKEYS]
+    #  Used in public key aggregation
+    pubkeys_serial = ['0x' + pubkey.hex() for pubkey in pubkeys]
+    case03_private_to_public_key = [
+        {
+            'input': int_to_hex(privkey),
+            'output': pubkey_serial,
+        }
+        for privkey, pubkey_serial in zip(PRIVKEYS, pubkeys_serial)
+    ]
+
+    case04_sign_messages = []
+    sigs = []  # used in verify
+    for privkey in PRIVKEYS:
+        for message in MESSAGES:
+            for domain in DOMAINS:
+                sig = bls.sign(message, privkey, domain)
+                case04_sign_messages.append({
+                    'input': {
+                        'privkey': int_to_hex(privkey),
+                        'message': '0x' + msg.hex(),
```

Comment:
This line has a bug and should be

`'message': '0x' + message.hex(),`

instead of

̀'message': '0x' + msg.hex(),`